pkg/api/customization/multiclusterapp/store.go

```diff
@@ -126,56 +135,107 @@ func (s *Store) setDisplayName(apiContext *types.APIContext, schema *types.Schem
+			rts, err := s.rtLister.List("", labels.NewSelector())
+			if err != nil {
+				return nil, err
 			}
-			clusterName := split[0]
-			projectName := split[1]
-			if !clustersOfTargetProjects[clusterName] {
-				// get roles from this cluster for this creator
-				crtbs, err := s.crtbLister.List(clusterName, labels.NewSelector())
-				if err != nil {
-					return nil, err
-				}
-				for _, crtb := range crtbs {
-					if crtb.UserName == callerID && !rolesToAddMap[crtb.RoleTemplateName] {
-						rolesToAddMap[crtb.RoleTemplateName] = true
-					}
+			for _, rt := range rts {
+				if rt.ClusterCreatorDefault && !rolesToAddMap[rt.Name] {
+					rolesToAddMap[rt.Name] = true
+				} else if rt.ProjectCreatorDefault && !rolesToAddMap[rt.Name] {
+					rolesToAddMap[rt.Name] = true
 				}
-				clustersOfTargetProjects[clusterName] = true
 			}
-			if !targetProjects[projectName] {
+			if len(rolesToAddMap) == 0 {
+				return nil, fmt.Errorf("Admin has passed no roles to multiclusterapp, and no cluster/project owner roles found")
+			}
+			rolesToAdd := make([]string, len(rolesToAddMap))
+			i := 0
+			for role := range rolesToAddMap {
+				rolesToAdd[i] = role
+				i++
+			}
+			values.PutValue(data, rolesToAdd, client.MultiClusterAppFieldRoles)
+		} else {
+			targetToRoles := make(map[string][]string)
+			for _, t := range targInterface {
+				projectRoleFound := false
+				isClusterCreator := false
+				target := convert.ToString(t[client.TargetFieldProjectID])
+				split := strings.SplitN(target, ":", 2)
+				if len(split) != 2 {
+					return nil, fmt.Errorf("invalid project name: %v", target)
+				}
+				clusterName := split[0]
+				projectName := split[1]
 				// get roles from this project for this creator
 				prtbs, err := s.prtbLister.List(projectName, labels.NewSelector())
 				if err != nil {
 					return nil, err
 				}
 				for _, prtb := range prtbs {
 					if prtb.UserName == callerID && !rolesToAddMap[prtb.RoleTemplateName] {
-						rolesToAddMap[prtb.RoleTemplateName] = true
+						projectRoleFound = true
+						if roles, ok := targetToRoles[target]; ok {
+							if !slice.ContainsString(roles, prtb.RoleTemplateName) {
+								targetToRoles[target] = append(roles, prtb.RoleTemplateName)
+							}
+						} else {
+							targetToRoles[target] = []string{prtb.RoleTemplateName}
+						}
 					}
 				}
-				targetProjects[projectName] = true
+
+				if !projectRoleFound {
+					// get roles from this cluster for this creator
+					crtbs, err := s.crtbLister.List(clusterName, labels.NewSelector())
+					if err != nil {
+						return nil, err
+					}
+					for _, crtb := range crtbs {
+						if crtb.UserName == callerID && !rolesToAddMap[crtb.RoleTemplateName] {
+							roleTemplate, err := s.rtLister.Get("", crtb.RoleTemplateName)
+							if err != nil {
+								return nil, err
+							}
+							if roleTemplate != nil && roleTemplate.ClusterCreatorDefault {
+								isClusterCreator = true
```

Comment:
This is the use case when user passed no explicit Roles, and so we're inheriting the user's roles in target projects. Now it could be the case that the creator is a cluster-owner, and should have access to all projects without having a prtb in those projects. So this check is done, when the creator does not have any roles in target project, to see whether the creator is an owner of the that project's cluster or not. And usually cluster-owner is the role with ClusterCreatorDefault. Does that sound okay to you @prachidamle ?